README.md
# sort
reads line of a file and sorts them in order


doc/changes
January 24, 1995	1.0 Initial effort                        (Jim Lynch)
May 12, 2003		1.1 Added NLS sort support            (Eric Auer, EA)
May 30, 2003		1.2 Improved i18n (KITTEN) usage                 (EA)
July 4, 2004		1.3 New languages. Enabled NLS sort by default,
                            replaced /N (enable) by /A (disable) switch  (EA)
July 9, 2004	        1.4 sort.exe/.com can sort max 395k/28k of data now
                            Old sort.com could only sort max 15k of data (EA)


source/sort.c
/*
*    SORT - reads line of a file and sorts them in order
*    Copyright  1995  Jim Lynch
*    Updated 2003-2004 by Eric Auer: NLS for sort order, smaller binary.
*
*    This program is free software; you can redistribute it and/or modify
*    it under the terms of the GNU General Public License as published by
*    the Free Software Foundation; either version 2 of the License, or
*    (at your option) any later version.
*
*    This program is distributed in the hope that it will be useful,
*    but WITHOUT ANY WARRANTY; without even the implied warranty of
*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*    GNU General Public License for more details.
*
*    You should have received a copy of the GNU General Public License
*    along with this program; if not, write to the Free Software
*    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "kitten.h"
#include <fcntl.h> /* O_RDONLY */
#define StdIN  0
#define StdOUT 1
#define StdERR 2

#ifndef __MSDOS__
#include <malloc.h>
#endif

#ifdef __MSDOS__
#include <dos.h> /* intdosx */

/*
NLS is int 21.65 -> get tables... country=-1 is current... The data:
02 - upcase table: 128 bytes of "what is that 0x80..0xff char in uppercase?"
03 - downcase table: 256 bytes of "what is that 0x00..0xff char in lowercase?"
  (only DOS 6.2+ with country.sys loaded, probably only codepage 866)
04 - filename uppercase table
05 - filename terminator table
06 - collating: 256 "values used to sort char 0x00..0xff" (I think you can
  use those byte values to "XLAT" (ASM) / look up (C) a number for each
  char, and then use that number for sorting comparisons)
07 - DBCS is somewhat different, see RBIL (double byte chars, DOS 4.0+)

The returned buffer contents are NOT the tables but a byte followed by
a far POINTER to the actual tables!
All tables are prefixed by a word indicating the table size (128, 256...)!
Minimum DOS version for NLS stuff is 3.3 ... For table request, give BX
and DX as codepage / country, or -1 to use current. You must provide a
buffer at ES:DI and tell its length in CX. On return, CY / buffer / CX set.
Only if NLSFUNC loaded, other codepages / countries than current are avail.

From DOS 4.0 on, you also have:
Upcase char:   21.6520.dl -> dl
Upcase array:  21.6521.ds:dx.cx (string at ds:dx, length cx)
Upcase string: 21.6522.ds:dx (string terminated by 0 byte)
*/

/* actually TINY was worst, because of non-malloc list[] array! */
#ifndef __TINY__		/* tiny model (.com) has no stack setting */
unsigned _stklen = 16383;	/* use bigger stack, for stdlib qsort...  */
#endif				/* default stack size is only 4096 bytes  */
/* for NORMAL (avg. 40 char/line) 10k line text, 6-8k of stack is ok */
/* on the other hand, 1400 "nnnn\tx" lines sorted /+5 take 32k stack */
/* (with 16k stack, you can "sort" up to 700 identical lines, etc.)  */

unsigned char far * collate;
#endif



#ifndef MAXPATH
#define MAXPATH 80
#endif

/* Optimizations 7/2004: sort bigger files both with COM and EXE version  -ea */
#ifdef __TINY__	/* .com: 8k+1k kitten, 7k*2 list array, ca. 28k max file size */
#define MAXRECORDS  7000	/* maximum number of records that can be
				 * sorted: we use an array of N pointers! */
#define MAXLEN	1023		/* maximum record length */
#else				/* if far pointers */
#define MAXRECORDS 12000	/* malloc limited to 64k -> max 16k items */
#define MAXLEN	8191		/* maximum record length */
#endif				/* !12k items means 48k for list[] alone! */
/* first 10k lines of Ralf Browns Interrupt List 61 are 382701 bytes, we */
/* can SORT them with "largest executable program size 626k (641,424)".  */
/* Limit for 626k seems to be - with 12k items and 12k stack - 400000 bytes */


int             rev;		/* reverse flag */
#ifdef __MSDOS__
int		nls;		/* NLS use flag */
#endif
int             help;		/* help flag */
int             sortcol;	/* sort column */
int             err = 0;	/* error counter */

void
WriteString(char *s, int handle) /* much smaller than fputs */
{
  write(handle, s, strlen(s));
}

int
cmpr(void *a, void *b)
{
    unsigned char *A, *B, *C;

    A = *(unsigned char **) a;
    B = *(unsigned char **) b;

    if (sortcol > 0) { /* "sort from column... " */
	if (strlen(A) > sortcol)
	    A += sortcol;
	else
	    A = "";
	if (strlen(B) > sortcol)
	    B += sortcol;
	else
	    B = "";
    }
    
    if (rev) { /* reverse sort: swap strings */
        /* (or swap sign of result, of course) */
        C = A;
        A = B;
        B = C;
    }

#ifdef __MSDOS__
    if (nls) {
        while (collate[A[0]] == collate[B[0]]) {
            /* we use collate in the while as well: two different *
             * bytes may have the same collate position...        */
            if (A[0] == '\0') return 0; /* both at end */
            A++;
            B++;
        }
        if (collate[A[0]] < collate[B[0]]) return -1;
        return 1;
    } else {
#endif
        return strcmp(A, B);
#ifdef __MSDOS__
    }
#endif
}

void
usage(nl_catd cat)
{
    if (cat != cat) {}; /* avoid unused argument error in kitten */
  
    WriteString("FreeDOS SORT v1.4\r\n", StdERR);	/* *** VERSION *** */

    if (err)
        WriteString(catgets(cat, 2, 0, "Invalid parameter\r\n"), StdERR);
    WriteString(catgets(cat, 0, 0, "    SORT [/R] [/+num] [/A] [/?] [file]\r\n"), StdERR);
    WriteString(catgets(cat, 0, 1, "    /R    Reverse order\r\n"), StdERR);
#ifdef __MSDOS__
#if 0
    WriteString(catgets(cat, 0, 2, "    /N    Enable NLS support\r\n"), StdERR);
#endif
    WriteString(catgets(cat, 0, 2, "    /A    Sort by ASCII, ignore COUNTRY\r\n"), StdERR);
#endif
    WriteString(catgets(cat, 0, 3, 
        "    /+num start sorting with column num, 1 based\r\n"), StdERR);
    WriteString(catgets(cat, 0, 4, "    /?    help\r\n"), StdERR);

}

int main(int argc, char **argv)
{
    char	filename[MAXPATH];
#ifdef __TINY__
    char	temp[MAXLEN + 1];
    char	*list[MAXRECORDS]; /* eats up lots of stack! */
#else
    char	*temp;
    char        **list;
#endif
    char	*cp;	/* option character pointer */
    int		nr;
    int		i;
    /* FILE	*fi; */	/* file descriptor */
    int		fi;	/* file HANDLE (fopen/... are big) */
    nl_catd	cat;	/* handle for localized messages (catalog) */

#ifdef __MSDOS__
    /* MAKE SURE THAT YOU USE BYTE ALIGNMENT HERE! */
    struct NLSBUF {
    	char id;
    	unsigned char far * content;
    } collbuf;
    union REGS     dosr;
    struct SREGS   doss;
#endif

    cat = catopen ("sort", 0);
  
    sortcol = 0;
    strcpy(filename, "");
    rev = 0;
#ifdef __MSDOS__
    nls = 1;	/* *** changed to default ON 7/2004 *** */
#endif

    while (--argc) {
	if (*(cp = *++argv) == '/') {
	    switch (cp[1]) {
	    case 'R':
	    case 'r':
		rev = 1;
		break;
#ifdef __MSDOS__
	    case 'N':
	    case 'n':
	    	nls = 1;	/* is the new default anyway ... */
		break;
	    case 'A':		/* use strict  ASCII  sort order */
	    case 'a':
	    	nls = 0;	/* new option 7/2004: suppress NLS collate */
		break;
#endif
	    case '?':
	    case 'h':
	    case 'H':
		help = 1;
		break;
	    case '+':
		sortcol = atoi(cp + 1);
		if (sortcol)
		    sortcol--;
		break;
	    default:
		err++;
	    }
	} else {		/* must be a file name */
	    strcpy(filename, *argv);
	}
    }

    if (err || help) {
	usage(cat);
	catclose(cat);
	exit(1);
    }

/* *** following was switch case "N" (in above parse) before 7/2004 *** */
#ifdef __MSDOS__
    if (nls) {
        if ( ((_osmajor >= 3) && (_osminor >= 3)) || (_osmajor > 3) ) {
            dosr.x.ax = 0x6506; /* get collate table */
            dosr.x.bx = 0xffff; /* default codepage  */
            dosr.x.dx = 0xffff; /* default country   */
            dosr.x.cx = 5;      /* buffer size   */
            doss.es = FP_SEG(&collbuf);
            dosr.x.di = FP_OFF(&collbuf);
            intdosx(&dosr,&dosr,&doss);
            if ((dosr.x.flags & 1) == 1) {
                WriteString(catgets(cat, 2, 1,
/* catgets ... */           "Error reading NLS collate table\r\n"),
                    StdERR);
                nls = 0;
            } else {
                /* ... CX is returned as table length ... */
                collate = collbuf.content; /* table pointer */
                collate++; /* skip leading word, which is */
                collate++; /* not part of the table */
 		/* "nls on" accepted now */
 		/* printf("NLS SORT ENABLED %d\r\n",dosr.x.cx); */
	    }
        } else {
           WriteString(catgets(cat, 2, 2,
/* catgets ... */       "Only DOS 3.3 or newer supports NLS!\r\n"),
               StdERR);
           nls = 0;	/* *** disable NLS sort order *** */
        }
    } else {
        /* accepting "nls off" is easy ;-) */
        /* printf("NLS SORT DISABLED AT USER REQUEST\r\n"); */
    }
#endif
/* *** end of former switch case "N" *** */

    fi = StdIN;			/* just in case */
    if (strlen(filename)) {
	if ((fi = open(filename, O_RDONLY)) == -1) {
	    /* was: ... fopen(...,"r") ... == NULL ... */
	    WriteString(catgets(cat, 2, 3, "SORT: Can't open "), StdERR);
	    WriteString(filename,StdERR);
	    WriteString(catgets(cat, 2, 4, " for read\r\n"), StdERR);
	    /* avoided 1.5k-2k overhead of *printf()... */
	    catclose(cat);
	    exit(2);
	}
    }
    
    (void)get_line(fi, NULL, 0); /* initialize buffers */

#ifndef __TINY__
    nr = -1;
    list = (char **) malloc(MAXRECORDS * sizeof(char *));
    if (list == NULL) goto out_of_mem;
    temp = (char *) malloc(MAXLEN + 1);
    if (temp == NULL) goto out_of_mem;
#endif
    
    for (nr = 0; nr < MAXRECORDS; nr++) {
	if (!get_line(fi, temp, MAXLEN))
	    /* was: fgets(temp, MAXLEN, fi) == NULL */
	    break;
	list[nr] = (char *) malloc(strlen(temp) + 1);
	    /* malloc might have big overhead, but we cannot avoid it */
	if (list[nr] == NULL) {
	    out_of_mem:
	    WriteString(catgets(cat, 2, 5, "SORT: Insufficient memory\r\n"),
	        StdERR);
#if 1
	    itoa(nr, temp, 10);
	    WriteString(temp, StdERR); /* print "offending" line number */
#endif
	    catclose(cat);
	    exit(3);
	}
	strcpy(list[nr], temp);
    }
#ifndef __TINY__
    /* free(temp); */	/* doesn't really help, qsort doesn't malloc anything */
    /* not using free() at all saves 0.7k (UPXed 0.3k) binary size... */
#endif

    if (nr == MAXRECORDS) {
	WriteString(catgets(cat, 2, 6,
            "SORT: number of records exceeds maximum\r\n"), /* catgets ... */
            StdERR);
	catclose(cat);
	exit(4);
    }

    qsort((void *) list, nr, sizeof(char *), cmpr);	/* *main action* */

    for (i = 0; i < nr; i++) {
	WriteString(list[i], StdOUT);
	WriteString("\r\n",StdOUT);
    }
    catclose(cat);
    return 0;
}



source/kitten.c
/* Functions that emulate UNIX catgets */

/* Copyright (C) 1999,2000,2001 Jim Hall <jhall@freedos.org> */
/* Kitten version 2003 by Tom Ehlert, heavily modified by Eric Auer 2003 */

/*
  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <stdio.h>			/* sprintf */
#include <stdlib.h>			/* getenv  */
#include <string.h>			/* strchr */
#include <fcntl.h>
#include <dos.h>

/* assert we are running in small model */
/* else pointer below has to be done correctly */
/* char verify_small_pointers[sizeof(void*) == 2 ? 1 : -1]; */

#include "kitten.h"

char catcontents[6144];	/* *** Maximum size of PROGRAM.%LANG% file *** */

struct catstring{
  char key1;
  char key2;
  char *text;
};

typedef struct catstring catstring_t;

catstring_t catpoints[128];		/* max 128 strings can be defined... */


/* Local prototypes */

int catread (char *catfile);		/* Reads a catfile into the hash */
char *processEscChars(char *line);  /* Converts c escape sequences to chars */

int get_char(int); /* not meant for external use    */
/* external use would cause consistency problems if */
/* value or related file of the file handle changes */

int mystrtoul(char *, int, int);


/* Globals */

nl_catd _kitten_catalog = 0;	/* _kitten_catalog descriptor or 0 */
char catfile[128];		/* full path to _kitten_catalog */

char getlbuf[1024]; /* read buffer (better speed) -- 7/2004: 1k, not 8k */
char * getlp;       /* current point in buffer       */
int getlrem = -1;   /* remaining bytes in buffer     */
char lastcr = 0;    /* for 2byte CR LF sequences     */



/* DOS handle based file usage */

int dos_open(char *filename, int mode)
{
	union REGS r;
        struct SREGS s;
	if (mode);		/* mode ignored - readonly supported */
	r.h.ah = 0x3d;
	r.h.al = 0;		/* read mode only supoported now !! */
	r.x.dx = FP_OFF(filename);
        s.ds = FP_SEG(filename);
	intdosx(&r,&r,&s);
	return ( (r.x.cflag) ? -1 : r.x.ax );
}


int dos_read(int file, void *ptr, unsigned count)
{
	union REGS r;
        struct SREGS s;
	r.h.ah = 0x3f;
	r.x.bx = file;
	r.x.cx = count;
	r.x.dx = FP_OFF(ptr);
        s.ds = FP_SEG(ptr);
	intdosx(&r,&r,&s);
	return ( (r.x.cflag) ? 0 : r.x.ax );
}


int dos_write(int file, void *ptr, unsigned count)
{
	union REGS r;
	struct SREGS s;
	r.h.ah = 0x40;
	r.x.bx = file;
	r.x.cx = count;
	r.x.dx = FP_OFF(ptr);
        s.ds = FP_SEG(ptr);
	intdosx(&r,&r,&s);
	return ( (r.x.cflag) ? 0 : r.x.ax );
}


void dos_close(int file)
{
	union REGS r;
	r.h.ah = 0x3e;
	r.x.bx = file;
	intdos(&r,&r);
}


/* Functions */

/**
 * On success, catgets() returns a pointer to an internal
 * buffer area containing the null-terminated message string.
 * On failure, catgets() returns the value 'message'.
 */

char * kittengets(int setnum, int msgnum, char * message)
{

  int i = 0;
  
  while ((catpoints[i].key1 != setnum) || (catpoints[i].key2 != msgnum)) {
    if ((catpoints[i].text == NULL) || (i>127)) /* at EOF */
      return message;
    i++;
  }

  if (catpoints[i].text == NULL)
    return message;
  else
    return (catpoints[i].text);
}


/**
 * Initialize kitten for program (name).
 */

nl_catd kittenopen(char *name)
{
  /* catopen() returns a message _kitten_catalog descriptor  *
   * of type nl_catd on success.  On failure, it returns -1. */

  char catlang[3];			/* from LANG environment var. */
  char *nlsptr;				/* ptr to NLSPATH */
  char *lang;                           /* ptr to LANG */
  int i;
  
  /* Open the _kitten_catalog file */
  /* The value of `_kitten_catalog' will be set based on catread */

  if (_kitten_catalog) { /* Already one open */
    write(1,"cat already open\r\n",strlen("cat already open\r\n"));
    return (-1);
  }

  for (i=0; i<128; i++)
    catpoints[i].text = NULL;

  if (strchr (name, '\\')) {
    /* unusual case: 'name' is a filename */
    write(1,"found \\\r\n",9);
    _kitten_catalog = catread (name);
    if (_kitten_catalog)
      return (_kitten_catalog);
  }

  /* If the message _kitten_catalog file name does not contain a directory *
   * separator, then we need to try to locate the message _kitten_catalog. */

  /* We will need the value of LANG, and may need a 2-letter abbrev of
     LANG later on, so get it now. */

  lang = getenv ("LANG");

  if (lang == NULL) {
      /* printf("no lang= found\n"); */ /* not fatal, though */
      /* Return failure - we won't be able to locate the cat file */
      return (-1);
  }

  if ( ( strlen(lang) < 2 ) ||
       ( (strlen(lang) > 2) && (lang[2] != '-') ) ) {
      /* Return failure - we won't be able to locate the cat file */
      return (-1);
  }

  memcpy(catlang, lang, 2);
  /* we copy the full LANG value or the part before "-" if "-" found */
  catlang[2] = '\0';

  /* step through NLSPATH */

  nlsptr = getenv ("NLSPATH");

  if (nlsptr == NULL) {
      /* printf("no NLSPATH= found\n"); */ /* not fatal either */
      /* Return failure - we won't be able to locate the cat file */
      return (-1);
   }

  catfile[0] = '\0';

  while (nlsptr != NULL) {
      char *tok = strchr(nlsptr, ';');
      int toklen;

      if (tok == NULL)
        toklen = strlen(nlsptr); /* last segment */
      else
        toklen = (int)(tok - nlsptr); /* segment terminated by ';' */
      
      /* catfile = malloc(toklen+1+strlen(name)+1+strlen(lang)+1); */
      /* Try to find the _kitten_catalog file in each path from NLSPATH */

      if ((toklen+6+strlen(name)) > sizeof(catfile)) {
        write(1,"NLSPATH overflow\r\n",strlen("NLSPATH overflow\r\n"));
        return 0; /* overflow in NLSPATH, should never happen */
      }

      /* Rule #1: %NLSPATH%\%LANG%\cat */

        memcpy(catfile, nlsptr, toklen);
        strcpy(catfile+toklen,"\\");
        strcat(catfile,catlang);
        strcat(catfile,"\\");
        strcat(catfile,name);
        _kitten_catalog = catread (catfile);
        if (_kitten_catalog)
	  return (_kitten_catalog);

      /* Rule #2: %NLSPATH%\cat.%LANG% */

        /* memcpy(catfile, nlsptr, toklen); */
        strcpy(catfile+toklen,"\\");
        strcat(catfile,name);
        strcat(catfile,".");
        strcat(catfile,catlang);
        _kitten_catalog = catread (catfile);
        if (_kitten_catalog)
	  return (_kitten_catalog);

      /* Grab next tok for the next while iteration */

        nlsptr = tok;
        if (nlsptr) nlsptr++;
      
    } /* while tok */

  /* We could not find it.  Return failure. */

  return (-1);
}


/**
 * Load a message catalog into memory.
 */

int catread (char *catfile)
{
  int   file;				/* pointer to the catfile */
  int   i;
  char  *where;
  char  *tok;

  /* Get the whole catfile into a buffer and parse it */
  
  file = open (catfile, O_RDONLY | O_TEXT);
  if (file < 0) 
      /* Cannot open the file.  Return failure */
      return 0;

  for (i=0; i<128; i++)
    catpoints[i].text = NULL;

  for (i=0; i<sizeof(catcontents); i++)
    catcontents[i] = '\0';

  /* Read the file into memory */
  i = read (file, catcontents, sizeof(catcontents)-1);

  if ((i == sizeof(catcontents)-1) || (i < 1))
      return 0; /* file was too big or too small */

  where = catcontents;
  i = 0; /* catpoints entry */

  do {
    char *msg;
    char *key;
    int key1 = 0;
    int key2 = 0;

    tok = strchr(where, '\n');

    if (tok == NULL) { /* done? */
      close(file);
      return 1; /* success */
    }

    tok[0] = '\0'; /* terminate here */ 
    tok--; /* guess: \r before \n */
    if (tok[0] != '\r')
      tok++; /* if not, go back */
    else {
      tok[0] = '\0'; /* terminate here already */
      tok++;
    }
    tok++; /* this is where the next line starts */

    if ( (where[0] >= '0') && (where[0] <= '9') &&
         ( (msg = strchr(where,':')) != NULL) ) {
      /* Skip everything which starts with # or with no digit */
      /* Entries look like "1.2:This is a message" */

      msg[0] = '\0'; /* remove : */
      msg++; /* go past the : */

      if ((key = strchr (where, '.')) != NULL)	{
        key[0] = '\0'; /* turn . into terminator */
        key++; /* go past the . */
        key1 = mystrtoul(where, 10, strlen(where));
        key2 = mystrtoul(key, 10, strlen(key));

        if ((key1 >= 0) && (key2 >= 0)) {
          catpoints[i].key1 = key1;
          catpoints[i].key2 = key2;
          catpoints[i].text = processEscChars(msg);
          if (catpoints[i].text == NULL) /* ESC parse error */
	    catpoints[i].text = msg;
          i++; /* next entry! */
        } /* valid keys */

      } /* . found */

    } /* : and digit found */

    where = tok; /* go to next line */

  } while (1);
}


void kittenclose (void)
{
  /* close a message _kitten_catalog */
  _kitten_catalog = 0;
}


/**
 * Parse a string that represents an unsigned integer.
 * Returns -1 if an error is found. The first size
 * chars of the string are parsed.
 */

int mystrtoul(char *src, int base, int size)
{
	int ret = 0;
	
	for (; size > 0; size--)  {
	  int digit;
	  int ch = *src;
          src++;

	  if (ch >= '0' && ch <= '9')
	    digit = ch - '0';
	  else if (ch >= 'A' && ch <= 'Z')
            digit = ch - 'A' + 10;
	  else if (ch >= 'a' && ch <= 'z')
            digit = ch - 'a' + 10;
	  else
	    return -1;

          if (digit >= base)
  	    return -1;
	  
	  ret = ret * base + digit;
	} /* for */

	return ret;
}		  	  	
	  	

/**
 * Process strings, converting \n, \t, \v, \b, \r, \f, \\,
 * \ddd, \xdd and \x0dd to actual chars.
 * (Note: \x is an extension to support hexadecimal)
 * This is used to allow the messages to use c escape sequences.
 * Modifies the line in-place (always same size or shorter).
 * Returns a pointer to input string.
 */

char *processEscChars(char *line)
{
  char *src = line;
  char *dst = line; /* possible as dst is shorter than src */

  /* used when converting \xdd and \ddd (hex or octal) characters */
  char ch;

  if (line == NULL)
    return line;

  /* cycle through copying characters, except when a \ is encountered. */
  while (*src != '\0') {
    ch = *src;
    src++;
    
    if (ch == '\\') {
      ch = *src; /* what follows slash? */
      src++;

      switch (ch) {
        case '\\': /* a single slash */
	  *dst = '\\';
	  dst++;
	  break;
        case 'n': /* a newline (linefeed) */
	  *dst = '\n';
	  dst++;
	  break;
        case 'r': /* a carriage return */
	  *dst = '\r';
	  dst++;
	  break;
        case 't': /* a horizontal tab */
	  *dst = '\t';
	  dst++;
	  break;
        case 'v': /* a vertical tab */
          *dst = '\v';
          dst++;
          break;
        case 'b': /* a backspace */
          *dst = '\b';
          dst++;
          break;
        case 'a': /* alert */
          *dst = '\a';
          dst++;
          break;
        case 'f': /* formfeed */
          *dst = '\f';
          dst++;
          break;
        case 'x': /* extension supporting hex numbers \xdd or \x0dd */
          {
            int chx = mystrtoul(src, 16, 2); /* get value */
            if (chx >= 0) { /* store character */
	       *dst = chx;
	       dst++;
               src += 2;
            } else /* error so just store x (loose slash) */
            {
              *dst = *src;
              dst++;
            }
          }
          break;
        default: /* just store letter (loose slash) or handle octal */
	  {
            int chx = mystrtoul(src, 8, 3); /* get value */
            if (chx >= 0) {/* store character */
	       *dst = chx;
	       dst++;
               src += 3;
            } else
            {
               *dst = *src;
               dst++;
            }
	  }
	  break;
      } /* switch */
    } /* if backslash */
      else
    {
      *dst = ch;
      dst++;
    }
  } /* while */

  /* ensure '\0' terminated */
  *dst = '\0';

  return line;
}


int get_char(int file) {
  int rval = -1;

  if (getlrem <= 0) { /* (re)init buffer */
    getlrem = read(file, getlbuf, sizeof(getlbuf));
    if (getlrem <= 0)
      return -1; /* fail: read error / EOF */
    getlp = getlbuf; /* init pointer */
  }

  if (getlrem > 0) { /* consume byte from buffer */
    rval = getlp[0];
    getlp++;
    getlrem--;
  }

  return rval;
}


/**
 * Read a line of text from file. You must call this with
 * a null buffer or null size to flush buffers when you are
 * done with a file before using it on the next file. Cannot
 * be used for 2 files at the same time.
 */

int get_line (int file, char *str, int size)
{
  int success = 0;
  int ch;

  if ((size == 0) || (str == NULL)) { /* re-init get_line buffers */
    getlp = getlbuf;
    getlrem = -1;
    lastcr = 0;
    return 0;
  }

  str[0] = '\0';

  while ( (size > 0) && (success == 0) ) {
    ch = get_char (file);
    if (ch < 0)
      break; /* (can cause fail if no \n found yet) */

    if (ch == '\r')
      ch = get_char (file); /* ignore \r */

    str[0] = ch;

    if ( (ch == '\n') || (ch == '\r') ) { /* done? */
      str[0] = '\0';
      return 1; /* success */
    }

    str++;
    size--;

  } /* while */

  str[0] = '\0'; /* terminate buffer */

  return success;

}

